mopidy_spotify/web.py

```diff
@@ -356,24 +358,114 @@ def __str__(self):
+        f"name,owner.id,type,uri,snapshot_id,tracks({TRACK_FIELDS}),"
+    )
+
     def __init__(self, client_id, client_secret, proxy_config):
-        self.user_name = None
-        self.user_country = None
         super(SpotifyOAuthClient, self).__init__(
             base_url="https://api.spotify.com/v1",
             refresh_url="https://auth.mopidy.com/spotify/token",
             client_id=client_id,
             client_secret=client_secret,
             proxy_config=proxy_config,
         )
+        self.user_id = None
+
+    def get_all(self, path, *args, **kwargs):
+        while path is not None:
+            logger.debug(f'Fetching page "{path}"')
+            result = self.get(path, *args, **kwargs)
+            path = result.get("next")
+            yield result
 
     def login(self):
-        user_profile = self.get("me")
-        if user_profile is None:
+        self.user_id = self.get("me").get("id")
+        if self.user_id is None:
             logger.error("Failed to load Spotify user profile")
             return False
         else:
-            self.user_name = user_profile.get("id")
-            self.user_country = user_profile.get("country")
-            logger.info(f"Logged into Spotify Web API as {self.user_name}")
+            logger.info(f"Logged into Spotify Web API as {self.user_id}")
             return True
+
+    def get_user_playlists(self, cache=None):
+        with utils.time_logger("get_user_playlists"):
+            pages = self.get_all(
+                "me/playlists", cache=cache, params={"limit": 50}
+            )
+            for page in pages:
+                for playlist in page.get("items", []):
+                    yield playlist
+
+    def get_playlist(self, uri, cache=None):
+        try:
+            parsed = parse_uri(uri)
+            if parsed.type != "playlist":
+                raise ValueError(
+                    f"Could not parse {repr(uri)} as a Spotify playlist URI"
+                )
+        except ValueError as exc:
+            logger.error(exc)
+            return {}
+
+        playlist = self.get(
+            f"playlists/{parsed.id}",
+            cache=cache,
+            params={"fields": self.PLAYLIST_FIELDS, "market": "from_token"},
+        )
+
+        tracks_path = playlist.get("tracks", {}).get("next")
+        track_pages = self.get_all(
+            tracks_path,
+            cache=cache,
+            params={"fields": self.TRACK_FIELDS, "market": "from_token"},
+        )
+
+        more_tracks = []
+        for page in track_pages:
+            more_tracks += page.get("items", [])
+        if more_tracks:
+            # Take a copy to avoid changing the cached response.
+            playlist = copy.deepcopy(playlist)
+            playlist.setdefault("tracks", {}).setdefault("items", [])
+            playlist["tracks"]["items"] += more_tracks
+
+        return playlist
+
+
+WebLink = collections.namedtuple("WebLink", ["uri", "type", "id", "owner"])
```

Comment:
Replace with a proper dataclass?

``` python
from dataclasses import dataclass
from typing import Optional

@dataclass
class WebLink:
    uri: str
    type: str  # This could also be an enum
    id: Optional[str] = None
    owner: Optional[str] = None
```

With this, you could also move `parse_uri()` to a classmethod `WebLink.from_uri(uri)`.